### ISSUE TITLE
mpd opus metadata is not shown on icecast

### ISSUE THREAD

**Issue by prietus** [MusicPlayerDaemon/MPD#830]
### Describe the bug
mpd track metadata using shout output and opus encoding is not shown on icecast 

## Expected Behavior
it should show tracks metadata

## Actual Behavior
no metadata at all

## Version

Music Player Daemon 0.21.22 (0.21.22)
Copyright 2003-2007 Warren Dukes <warren.dukes@gmail.com>
Copyright 2008-2018 Max Kellermann <max.kellermann@gmail.com>
This is free software; see the source for copying conditions.  There is NO
warranty; not even MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Database plugins:
 simple proxy upnp

Storage plugins:
 local smbclient udisks nfs curl

Neighbor plugins:
 smbclient upnp udisks

Decoders plugins:
 [mad] mp3 mp2
 [mpg123] mp3
 [vorbis] ogg oga
 [oggflac] ogg oga
 [flac] flac
 [opus] opus ogg oga
 [sndfile] wav aiff aif au snd paf iff svx sf voc w64 pvf xi htk caf sd2
 [audiofile] wav au aiff aif
 [dsdiff] dff
 [dsf] dsf
 [hybrid_dsd] m4a
 [faad] aac
 [mpcdec] mpc
 [wavpack] wv
 [modplug] 669 amf ams dbm dfm dsm far it med mdl mod mtm mt2 okt s3m stm ult umx xm
 [mikmod] amf dsm far gdm imf it med mod mtm s3m stm stx ult uni xm
 [wildmidi] mid
 [fluidsynth] mid
 [ffmpeg] 16sv 3g2 3gp 4xm 8svx aa3 aac ac3 adx afc aif aifc aiff al alaw amr anim apc ape asf atrac au aud avi avm2 avs bap bfi c93 cak cin cmv cpk daud dct divx dts dv dvd dxa eac3 film flac flc fli fll flx flv g726 gsm gxf iss m1v m2v m2t m2ts m4a m4b m4v mad mj2 mjpeg mjpg mka mkv mlp mm mmf mov mp+ mp1 mp2 mp3 mp4 mpc mpeg mpg mpga mpp mpu mve mvi mxf nc nsv nut nuv oga ogm ogv ogx oma ogg omg opus psp pva qcp qt r3d ra ram rl2 rm rmvb roq rpl rvc shn smk snd sol son spx str swf tak tgi tgq tgv thp ts tsp tta xa xvid uv uv2 vb vid vob voc vp6 vmd wav webm wma wmv wsaud wsvga wv wve
 [gme] ay gbs gym hes kss nsf nsfe sap spc vgm vgz
 [pcm]

Filters:
 libsamplerate soxr

Tag plugins:
 id3tag

Output plugins:
 shout null fifo pipe alsa ao oss openal solaris pulse jack httpd recorder

Encoder plugins:
 null vorbis opus lame twolame wave flac

Archive plugins:
 [bz2] bz2
 [zzip] zip
 [iso] iso

Input plugins:
 file archive alsa tidal qobuz curl ffmpeg smbclient nfs mms cdio_paranoia

Playlist plugins:
 extm3u m3u pls xspf asx rss soundcloud flac cue embcue

Protocols:
 file:// alsa:// tidal:// qobuz:// http:// https:// gopher:// rtp:// rtsp:// rtmp:// rtmpt:// rtmps:// smb:// nfs:// mms:// mmsh:// mmst:// mmsu:// cdda://

Other features:
 avahi dbus udisks epoll icu inotify ipv6 systemd tcp un

## Log

from icecast:
[2020-04-14  22:35:20] INFO admin/admin_handle_request Received admin command metadata on mount '/mpd.opus'
[2020-04-14  22:35:20] EROR admin/command_metadata Got legacy SOURCE-style metadata update command on source connected with PUT at mountpoint /mpd.opus
[2020-04-14  22:35:20] EROR admin/command_metadata Got legacy shoutcast-style metadata update command on source that does not accept it at mountpoint /mpd.opus

**Comment by MaxKellermann** [comment]
What do those Icecast log messages mean?
What is your MPD configuration?

**Comment by prietus** [comment]
mpd.conf

music_directory         "~/Música"
playlist_directory              "~/Música/playlists"
log_file                        "~/.config/mpd/log"
pid_file                        "~/.config/mpd/pid"
state_file                      "~/.config/mpd/state"
sticker_file                    "~/.config/mpd/sticker.sql"
bind_to_address         "any"
port                            "6600"
log_level                       "default"
restore_paused "no"
auto_update     "yes"
auto_update_depth "3"
follow_outside_symlinks "yes"
follow_inside_symlinks          "yes"
zeroconf_enabled                "yes"
zeroconf_name                   "Music Player Daemon"

input {
plugin "tidal"
  token   ""
  username ""
  password ""
 audioquality "HI_RES"
}

audio_output {
    type        "shout"
    encoding    "opus"
protocol "icecast2"
    name        "shout"
    host        "19.10.10.10"
    port        "8000"
    mount       "/mpd.opus"
    password    ""
   quality     "5.0"
    format      "48000:24:2"
    user        "source"
}


audio_output {
        type            "pulse"
        name            "pulseaudio"
        format          "384000:32:2"
}

**Comment by prietus** [comment]
so it looks like this is when you try to use the metadata update endpoint to update opus metadata
but this is not supperted. it looks like is using legacy source style metadata

**Comment by MaxKellermann** [comment]
What exactly is not supported?
What is the "metadata update endpoint"?
And what is "legacy source style metadata"?

**Comment by prietus** [comment]
TL;DR
track metadata is not shown on icecast while streaming from mpd using opuc codec.
what is codec? what is metadata? what is mpd....?

**Comment by prietus** [comment]
the endpoint is only for ICY streams. that is MP3 and AAC.  
the correct way is sending vorbis comments as per Opus specs.
or just 'speak Opus as per Opus specs'.

**Comment by prietus** [comment]
is that ok?

**Comment by ePirat** [comment]
Hi, I am one of the Icecast devs and had a look at the implementation in MPD.
The relevant part for this issue is here:

https://github.com/MusicPlayerDaemon/MPD/blob/a2f5a63bbc1a21be672d358a107a3a206c0c5369/src/output/plugins/ShoutOutputPlugin.cxx#L366-L392

So it seems MPD already does the right thing if the encoder supports stream tags, if I read the code correctly. And only if the encoder lacks this, falls back to the legacy ICY metadata using the update endpoint.

The Opus encoder seems to support this:

https://github.com/MusicPlayerDaemon/MPD/blob/256cfc545ddf05ed72784922f3ec038da6747032/src/encoder/plugins/OpusEncoderPlugin.cxx#L383

So I guess maybe the MPD version that @prietus is using is too old and does not has this yet?

**Comment by MaxKellermann** [comment]
> So I guess maybe the MPD version that @prietus is using is too old and does not has this yet?

Thanks @ePirat for coming over to help!
No, the Opus encoder plugin needs to have OpusTags explicitly enabled in configuration
https://www.musicpd.org/doc/html/plugins.html#id23
This is because there were big problems with the initial Opus encoder plugin, because many players didn't expect OpusTags packets in the middle of a stream.

**Comment by MaxKellermann** [comment]
But anyway, why does this problem exist, when there are Icy tags, and Icecast recognizes them and even logs their availability?
We now know how to make Icecast happier by enabling OpusTags, but why does Icecast recognize "legacy" Icy tags, but then ignores them?

**Comment by ePirat** [comment]
ICY is an awful hack we do not support for anything other than "legacy" formats that do not support proper metadata, we never supported them for any Icecast-supported format that can have "proper" metadata muxed into the stream.

What Icecast did for supported formats in the past was to [mux the metadata from the legacy endpoint into the stream][1], like the source client should do, but this leads to sometimes surprising behavior of "late" and incomplete metadata as the legacy endpoint was never intended for this and it's the source client that should do the muxing. Only then its ensured timing of metadata is correct vs. sending it out of band. So we did not implement it for any new formats we added support for.

it's true that some players had initially problems with chained Ogg but I believe most have fixed their issues and while this is an annoying problem, it's not really something we want to work around by using ancient hacks like the ICY stuff.

[1]: https://git.xiph.org/?p=icecast-server.git;a=blob;f=src/format_vorbis.c;h=c3bf7bc4db7f86c787811f198bca050cb8d544d8;hb=HEAD#l413

**Comment by MaxKellermann** [comment]
> we never supported them for any Icecast-supported format that can have "proper" metadata muxed into the stream

That is surprising, because MPD uses libshout, which is Icecast's official library for interfacing with Icecast, and its documentation does not state that `shout_metadata_add()` is legacy, obsolete, deprecated or anything.

Anyway, this bug report is bogus, because the problem is that Icecast refuses to accept metadata from MPD, and MPD can be configured in a way to send OpusTags instead of libshout-based tags.

**Comment by ePirat** [comment]
@MaxKellermann We do say that, at least in the [git master version of the documentation][1]:

> These functions currently only make sense for MP3 streams. Vorbis streams are expected
to embed metadata as vorbis comments in the audio stream.

[1]: https://git.xiph.org/?p=icecast-libshout.git;a=blob;f=doc/libshout.xml;h=20ea926b9f30a8865f5290af33bcbc361f2d3df8;hb=HEAD#l892

**Comment by MaxKellermann** [comment]
I've never seen this obscure XML document. It is not shipped on Debian, and I don't like reading XML.
The C headers have API documentation, and that's what I read. Those headers have deprecation warnings for old functions, and I would expect the same for `shout_metadata_add()`.

**Comment by ePirat** [comment]
This XML the docbook thing the documentation is generated from, it's not supposed to be read but generated to HTML or similar and should be included in generated form when shipped.

We unfortunately can not just add deprecation warnings for things like `shout_metadata_add` as those are still needed for MP3 as there is really no other way to handle metadata unfortunately. So it would cause warnings in code which would be unfixable as there is no other way to do it.

**Comment by MaxKellermann** [comment]
I didn't mean "deprecated" attributes, but text explaining the deprecation, and Icecast's refusal to accept `shout_metadata_add` data for anything but MP3. Not even the XML document mentions this - it only says "is expected". Who expects? What if that expectation fails?

**Comment by ePirat** [comment]
I agree it should be explained better in the documentation. I'll see if I can improve that for the next libshout release!
